### PR TITLE
feat(lint): add default_iter_empty

### LINKED ISSUE #8915: Prefer std::iter::empty() to std::iter::Empty::default()
### What it does

Replace `std::iter::Empty::default()` with `std::iter::empty()`.

These two are equivalent, but one should be preferred: `iter::empty` requires learning less of the API than `iter::Empty::default` and it's also simpler.

I actually made this mistake consistently in a PR, it seems silly, but it happens! I caught this during code review and thought, hey, clippy should be able to detect that.

### Lint Name

iter-empty-not-iter-empty-default

### Category

style

### Advantage

- It's shorter
- Rust generally discourages you from knowing the concrete type of an iterator combinator, e.g. you shouldn't really need to know Map or Empty, instead you should just call the map() and empty() functions. If you need to name the output type you should use `impl Iterator` instead.

### Drawbacks

_No response_

### Example

```rust
let x = std::iter::Empty::default();
```

Could be written as:

```rust
let x = std::iter::empty();
```

### PR DESCRIPTION
**Pull request by kyoto7250**
close #8915 

This PR adds `default_iter_empty` lint.

This lint checks `std::iter::Empty::default()` and replace with `std::iter::empty()`.

Thank you in advance.

---

changelog: add `default_instead_of_iter_empty` lint.

### CODE REVIEW

File: clippy_lints/src/default_iter_empty.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let _ = std::iter::empty::<usize>();
+    /// let iter: std::iter::Empty<usize> = std::iter::empty();
+    /// ```
+    #[clippy::version = "1.63.0"]
+    pub DEFAULT_ITER_EMPTY,
+    style,
+    "Check `std::iter::Empty::default()` and replace with `std::iter::empty()`"
```
[the remark this answers is not in the document]
Comment (kyoto7250) [reply clippy_lints/src/default_iter_empty.rs:29]: @xFrednet 
I will commit, thank you!

File: clippy_lints/src/default_iter_empty.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    /// let _ = std::iter::Empty::<usize>::default();
+    /// let iter: std::iter::Empty<usize> = std::iter::Empty::default();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let _ = std::iter::empty::<usize>();
+    /// let iter: std::iter::Empty<usize> = std::iter::empty();
+    /// ```
+    #[clippy::version = "1.63.0"]
+    pub DEFAULT_ITER_EMPTY,
```
Comment (Alexendoo) [inline clippy_lints/src/default_iter_empty.rs:27]: For the lint name `DEFAULT_INSTEAD_OF_ITER_EMPTY` would be more specific

File: clippy_lints/src/default_iter_empty.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                span_lint_and_sugg(
+                    cx,
+                    DEFAULT_ITER_EMPTY,
+                    expr.span,
+                    "`std::iter::empty()` is shorter than `std::iter::Empty::default()` and same functionality",
+                    "consider calling `std::iter::empty()` instead",
+                    sugg,
+                    applicability
+                );
+        }
```
Comment (Alexendoo) [inline clippy_lints/src/default_iter_empty.rs:52]: style nit:
```suggestion
            && match_def_path(cx, *def_id, &paths::ITER_EMPTY)
        {
            let mut applicability = Applicability::MachineApplicable;
            let sugg = make_sugg(cx, ty_path, &mut applicability);
            span_lint_and_sugg(
                cx,
                DEFAULT_ITER_EMPTY,
                expr.span,
                "`std::iter::empty()` is shorter than `std::iter::Empty::default()` and same functionality",
                "consider calling `std::iter::empty()` instead",
                sugg,
                applicability,
            );
        }
```

File: clippy_lints/src/default_iter_empty.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+fn make_sugg(cx: &LateContext<'_>, ty_path: &rustc_hir::QPath<'_>, applicability: &mut Applicability) -> String {
+    if let Some(last) = last_path_segment(ty_path).args
+        && let Some(iter_ty) = last.args.iter().find_map(|arg| match arg {
+            GenericArg::Type(ty) => Some(ty),
+            _ => None,
+        }) {
+            format!("std::iter::empty::<{}>()", snippet_with_applicability(cx, iter_ty.span, "..", applicability))
+        } else {
+            "std::iter::empty()".to_owned()
+        }
```
Comment (Alexendoo) [inline clippy_lints/src/default_iter_empty.rs:65]:
```suggestion
        })
    {
        format!("std::iter::empty::<{}>()", snippet_with_applicability(cx, iter_ty.span, "..", applicability))
    } else {
        "std::iter::empty()".to_owned()
    }
```

File: clippy_lints/src/default_iter_empty.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            && let QPath::Resolved(None, path) = ty_path
+            && let def::Res::Def(_, def_id) = &path.res
+            && match_def_path(cx, *def_id, &paths::ITER_EMPTY) {
+                let mut applicability = Applicability::MachineApplicable;
+                let sugg = make_sugg(cx, ty_path, &mut applicability);
+                span_lint_and_sugg(
+                    cx,
+                    DEFAULT_ITER_EMPTY,
+                    expr.span,
+                    "`std::iter::empty()` is shorter than `std::iter::Empty::default()` and same functionality",
```
Comment (Alexendoo) [inline clippy_lints/src/default_iter_empty.rs:47]: `Empty::default()` isn't that much longer than `iter::empty()`, I'd say instead mention that it's the more typical way to create the struct

File: clippy_lints/src/default_iter_empty.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    "consider calling `std::iter::empty()` instead",
```
Comment (Alexendoo) [inline clippy_lints/src/default_iter_empty.rs:48]: This makes the suggestion repeat `std::iter::empty` twice, the message can be something short like "try"/"replace with"

File: tests/ui/default_iter_empty.rs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+// run-rustfix
+#![warn(clippy::default_iter_empty)]
+use std::collections::HashMap;
+
+fn main() {
+    // Do lint.
+    let _ = std::iter::Empty::<usize>::default();
+    let _ = std::iter::Empty::<HashMap<usize, usize>>::default();
+    let _foo: std::iter::Empty<usize> = std::iter::Empty::default();
```
Comment (Alexendoo) [inline tests/ui/default_iter_empty.rs:9]: Could do with some tests with `use std::iter::Empty`, maybe some for `Default::default()` (probably shouldn't lint)

File: clippy_lints/src/default_iter_empty.rs (hunk 2 above)
Comment (kyoto7250) [reply clippy_lints/src/default_iter_empty.rs]: I fixed in https://github.com/rust-lang/rust-clippy/pull/8989/commits/ea1c1a5646330105fb2d656f9354ad16de19fe9d

File: clippy_lints/src/default_iter_empty.rs (hunk 5 above)
Comment (kyoto7250) [reply clippy_lints/src/default_iter_empty.rs]: I fixed https://github.com/rust-lang/rust-clippy/pull/8989/commits/2b6232c5e78e5c171b341459e84ed9923a697cf1

File: tests/ui/default_iter_empty.rs (hunk 7 above)
Comment (kyoto7250) [reply tests/ui/default_iter_empty.rs]: I fixed in https://github.com/rust-lang/rust-clippy/pull/8989/commits/9fd048e744d57b338565defd235fbfd9f9d386fd

File: clippy_lints/src/default_iter_empty.rs (hunk 6 above)
Comment (kyoto7250) [reply clippy_lints/src/default_iter_empty.rs]: I fixed in https://github.com/rust-lang/rust-clippy/pull/8989/commits/2b6232c5e78e5c171b341459e84ed9923a697cf1

File: clippy_lints/src/renamed_lints.rs [hunk 8]
```diff
@@ -7,6 +7,7 @@ pub static RENAMED_LINTS: &[(&str, &str)] = &[
     ("clippy::box_vec", "clippy::box_collection"),
     ("clippy::const_static_lifetime", "clippy::redundant_static_lifetimes"),
     ("clippy::cyclomatic_complexity", "clippy::cognitive_complexity"),
+    ("clippy::default_iter_empty", "clippy::default_instead_of_iter_empty"),
```
[the remark this answers is not in the document]
Comment (kyoto7250) [reply clippy_lints/src/renamed_lints.rs]: I didn't know that, thanks!

File: clippy_lints/src/default_instead_of_iter_empty.rs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            && let def::Res::Def(_, def_id) = &path.res
+            && match_def_path(cx, *def_id, &paths::ITER_EMPTY)
+        {
+            let mut applicability = Applicability::MachineApplicable;
+            let sugg = make_sugg(cx, ty_path, &mut applicability);
+            span_lint_and_sugg(
+                cx,
+                DEFAULT_INSTEAD_OF_ITER_EMPTY,
+                expr.span,
+                "`std::iter::empty()` is shorter than `std::iter::Empty::default()` and same functionality",
```
[the remark this answers is not in the document]
Comment (kyoto7250) [reply clippy_lints/src/default_instead_of_iter_empty.rs]: I fixed in [01a4ecf](https://github.com/rust-lang/rust-clippy/pull/8989/commits/01a4ecfcc7b40fd0e8fd711142a88358ea5d4dd7)

**Comment by kyoto7250** [comment]
I pushed some commits.

If we need to do `git squash`, please let me know.

**Comment by Alexendoo** [comment]
Yeah a squash would be good thanks

**Review by Alexendoo — APPROVED** [review]

**Comment by Alexendoo** [comment]
Thanks!

@bors r+

**Pull request merged**